File: python/aoc_2023/day01.py
import re
from dataclasses import dataclass
from functools import lru_cache

maps = {'one': '1',
        'two' : '2',
        'three': '3',
        'four': '4',
        'five': '5',
        'six': '6',
        'seven': '7',
        'eight': '8',
        'nine': '9'}

@lru_cache
def numbers_pattern():
    combined_pattern = '|'.join(maps.keys()) + r"|\d"
    pattern = re.compile(combined_pattern)
    return pattern

@dataclass
class Calibration:
    raw: str

    @staticmethod
    def from_line(line: str):
        return Calibration(line)
    
    def value(self) -> int:
        "Return the part-1 digit extraction (characters that are digits only)"
        digits = [c for c in self.raw if c.isdigit()]
        return int(digits[0] + digits[-1])
    
    def value2(self) -> int:
        "Return the part-2 digit extraction (spelled out or actual digits)"
        pattern = numbers_pattern()
        # Need to do this loop because some numbers have overlapping characters (eightwo -> 8,2)
        matches = []
        for i in range(len(self.raw)):
            match = pattern.match(self.raw, i)
            if match:
                # match could be a digit or a word. If it's a word, convert to digit with map lookup
                word_or_digit = match.group()
                if word_or_digit in maps:
                    matches.append(maps[word_or_digit])
                else:
                    matches.append(word_or_digit)
        return int(matches[0] + matches[-1])

File: rust/src/day01.rs
use regex::Regex;
use std::collections::HashMap;

// lazy_static is kind of like lru_cache in that it'll only run this code once
lazy_static::lazy_static! {
    static ref MAPS: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();
        m.insert("one", "1");
        m.insert("two", "2");
        m.insert("three", "3");
        m.insert("four", "4");
        m.insert("five", "5");
        m.insert("six", "6");
        m.insert("seven", "7");
        m.insert("eight", "8");
        m.insert("nine", "9");
        m
    };
    static ref NUMBERS_PATTERN: Regex = {
        let pattern = MAPS.keys().cloned().collect::<Vec<_>>().join("|") + r"|\d";
        Regex::new(&pattern).unwrap()
    };
}

pub struct Calibration {
    raw: String,
}

impl Calibration {
    pub fn from_line(line: &str) -> Calibration {
        Calibration {
            raw: line.to_string(),
        }
    }

    pub fn value(&self) -> i32 {
        let digits: Vec<char> = self.raw.chars().filter(|c| c.is_digit(10)).collect();
        let first = digits.first().unwrap();
        let last = digits.last().unwrap();
        let combined = format!("{}{}", first, last);
        combined.parse::<i32>().unwrap()
    }

    pub fn value2(&self) -> i32 {
        let mut digits = Vec::new();
        let mut start = 0;
        while let Some(matched) = NUMBERS_PATTERN.find(&self.raw[start..]) {
            let word_or_digit = matched.as_str();
            start += matched.start() + 1;

            let map_value = match MAPS.get(word_or_digit) {
                Some(v) => v,
                None => word_or_digit,
            };
            digits.push(map_value);
        }
        let first = digits.first().unwrap();
        let last = digits.last().unwrap();
        let combined = format!("{}{}", first, last);
        combined.parse::<i32>().unwrap()
    }
}
